library/ssl_tls.c

```diff
@@ -5818,6 +5863,56 @@ void mbedtls_ssl_handshake_free( mbedtls_ssl_context *ssl )
 
     mbedtls_platform_zeroize( handshake,
                               sizeof( mbedtls_ssl_handshake_params ) );
+
+#if defined(MBEDTLS_SSL_VARIABLE_BUFFER_LENGTH)
+    /* If the buffers are too big - reallocate */
+    int modified = 0;
+    unsigned char* resized_buffer = NULL;
+    uint32_t buf_len = mbedtls_ssl_get_input_buflen( ssl );
+
+    if( ssl->in_buf != NULL && ssl->in_buf_len > buf_len && ssl->in_left < buf_len )
+    {
+        resized_buffer = mbedtls_calloc( 1,  buf_len );
+        if( resized_buffer != NULL )
+        {
+            memcpy( resized_buffer, ssl->in_buf, buf_len );
+            mbedtls_platform_zeroize( ssl->in_buf, ssl->in_buf_len );
+            mbedtls_free( ssl->in_buf );
+
+            ssl->in_buf = resized_buffer;
+            ssl->in_buf_len = buf_len;
+            modified = 1;
+        }
+        else
+        {
+           MBEDTLS_SSL_DEBUG_MSG( 1, ( "output buffer resizing failed - out of memory" ) );
+        }
+    }
+
+    buf_len = mbedtls_ssl_get_output_buflen( ssl );
+    if( ssl->out_buf != NULL && ssl->out_buf_len > mbedtls_ssl_get_output_buflen( ssl ) && ssl->out_left < buf_len )
+    {
+        resized_buffer = mbedtls_calloc( 1, buf_len );
+        if( resized_buffer != NULL )
+        {
+            memcpy( resized_buffer, ssl->out_buf, buf_len );
+            mbedtls_platform_zeroize( ssl->out_buf, ssl->out_buf_len );
+            mbedtls_free( ssl->out_buf );
+
+            ssl->out_buf = resized_buffer;
+            ssl->out_buf_len = buf_len;
+            modified = 1;
+        }
+        else
+        {
+           MBEDTLS_SSL_DEBUG_MSG( 1, ( "output buffer resizing failed - out of memory" ) );
+        }
+    }
+    if( modified )
+    {
+        mbedtls_ssl_reset_in_out_pointers( ssl );
```

Comment:
Why is it important to wait until here to call mbedtls_ssl_reset_in_out_pointers()? Add a C comment to explain the efficiency gained by waiting until here.

Reply:
It's just a small improvement to not call `mbedtls_ssl_reset_in_out_pointers` twice, since the changed values won't be used between the first and second buffer resizes. It seems pretty simple, how would you like me to explain it?
`Update pointers here to avoid doing it twice.`?